test/EndToEndTests/Microsoft.OData.E2E.Portable/AsynchronousTests/AsynchronousQueryTests.cs

```diff
@@ -1197,6 +1197,36 @@ public void Linq_Where_GreaterThan()
             this.WaitForTestToComplete();
         }
 
+        /// <summary>
+        /// When DataServiceContext.MakeIdPredicateByKey=false, An expression with an Id as a predicate should create an Uri with a $filter in query options
+        /// </summary>
+        [Fact]
+        public void Linq_Where_Generates_Filter_By_Id_When_MakeIdPredicateByKey_Is_False()
+        {
+            var context = this.CreateWrappedContext<DefaultContainer>().Context;
+            context.MakeIdPredicateByKey = false;
+            var query = (from c in context.Customer
+                         where c.CustomerId == -10
+                         select new Customer { Name = c.Name, CustomerId = c.CustomerId }) as DataServiceQuery<Customer>;
+            var uri = query.ToString();
+            Assert.Contains("$filter=CustomerId eq -10", uri);
+        }
+
+        /// <summary>
+        /// When DataServiceContext.MakeIdPredicateByKey=true, An expression with an Id as a predicate should create a ByKey Uri
+        /// </summary>
+        [Fact]
+        public void Linq_Where_Generates_Filter_By_Id_When_MakeIdPredicateByKey_Is_True()
+        {
+            var context = this.CreateWrappedContext<DefaultContainer>().Context;
+            context.MakeIdPredicateByKey = false;
+            var query = (from c in context.Customer
+                         where c.CustomerId == -10
+                         select new Customer { Name = c.Name, CustomerId = c.CustomerId }) as DataServiceQuery<Customer>;
+            var uri = query.ToString();
+            Assert.Contains("Customer(-10)", uri);
+        }

```

Comment:
Judging from the method name and the comment, the intention was to set it true `context.MakeIdPredicateByKey` to true.

However, the fact that this test is passing regardless indicates that there's probably a bug. Maybe the uri contains both the "Customer(-10)" segment and the "$filter=Id eq -10" section. Just to be safe, I think your assertion should check for full uri equality instead of just checking whether the id segment is contained. Or at least specifically assert that it does not contain contain the filter section. But I think the full url check would be safer.

Reply:
@habbes I don't think there's any problem with this test. It does confirm that when the value of `MakeIdPredicateKey` is `false`, the key is included in the resource path: `Customer(-10)` while the previous test demonstrated that when the value if `true`, the key field appears on the filter: `$filter=CustomerId eq -10`

Reply:
@gathogojr @KenitoInc  I may have gotten things mixed up. But looking at the updated code now, I think I'm still confused. The option `MakeIdPredicateByKey` is supposed to result in `Customer(-10)` when it's set to `true`. However, he had set it to `false` and it still contained `Customer(-10)`. Now looking at the updated test, he has set it to `true` and it still contains `Customer(-10)`. It's as if the key segment is in the url regardless of the value of `MakeIdPredicateByKey`.

Also regarding your statement

> while the previous test demonstrated that when the value if true, the key field appears on the filter: $filter=CustomerId eq -10

Note that in the previous, the `MakeIdPredicateKey` is actually set to `false`, not `true` as you state. I could be wrong, but it may even be possible that it's setting both the `$filter` option and the key segment regardless of the value of the setting.

Just to be sure, I propose that @KenitoInc modify both tests such that they perform an `Assert.Equal()` instead of `Assert.Contains()`. In any case this wouldn't hurt, and it's more faithful to the intention since the current tests don't verify whether the filter or key is excluded from the url when the other is included.

Reply:
#Resolved